### ISSUE TITLE
Creating new data loaders, transformations and/or data exporters | Documentation

### ISSUE THREAD

**Issue by return007** [mage-ai/mage-ai#1255]
First things first, what a beauty Mage is ❤️!  

I would like to add some generic 3rd party data loaders and exporters, so I'm looking for some documentation on how to add one. So far, I went through existing documentation on Blocks (https://github.com/mage-ai/mage-ai/blob/master/docs/blocks/data_loading.md) and Contributing (https://github.com/mage-ai/mage-ai/tree/master/docs/contributing) and eyeballed some code (especially [mage_integrations](https://github.com/mage-ai/mage-ai/tree/master/mage_integrations/mage_integrations)) but couldn't get much idea how it would come up in UI, etc. 

Do we have some documentation already available on this?

**Comment by tommydangerous** [comment]
Thank you so much for the compliments and support!

I'll share the docs for adding a custom source.

TLDR: There are 2 ways to add
1. Via Data integration pipeline
2. Via Template for batch pipeline

**Comment by tommydangerous** [comment]
Data integration overview: https://github.com/mage-ai/mage-ai/blob/master/docs/data_integrations/README.md

How to create a data integration pipeline: https://github.com/mage-ai/mage-ai/blob/master/docs/guides/pipelines/DataIntegrationPipeline.md

Sources: https://github.com/mage-ai/mage-ai/blob/master/docs/data_integrations/README.md

How to add a source: https://github.com/mage-ai/mage-ai/blob/master/docs/guides/data_integrations/sources/AddNewSource.md

**Comment by tommydangerous** [comment]
Please ping here for any questions and we'll answer ASAP.

Or, live chat with all of us in Slack: https://www.mage.ai/chat

**Comment by tommydangerous** [comment]
If you want to add it to be used by the batch pipeline, here is an example of an S3 template: https://github.com/mage-ai/mage-ai/blob/master/mage_ai/data_preparation/templates/data_loaders/s3.py

**Comment by return007** [comment]
Thanks @tommydangerous . That was quick! 

I'll explore these.  Feel free to close the issue. If there's anything more, I'll catch-up on Slack ☺️

**Comment by tommydangerous** [comment]
We're extremely responsive!

Feel free to ping me/us anytime.

Also just want to clarify the difference of adding 3rd party templates/sources:

When developers want to write a lot of their own custom data pipeline, they use a batch pipeline. They can add as many blocks of code as they want. In those blocks, they write Python, SQL, or R. They are responsible for writing the logic to fetch data or export data from wherever source they want. That's where templates for data loaders and data exporters come in handy.

If a developer doesn't want to write code, they use a data integration pipeline, configure their pipeline with a few clicks and YAML file entries, and bam all done.

Depending on which you want to use will determine which code you want to follow.

**Comment by return007** [comment]
Got it, thanks for clarification.

**Comment by tommydangerous** [comment]
@return007 just checking in, were you able to get all the help you needed?
